memcopy.h

```diff
@@ -342,53 +291,108 @@ static inline unsigned char *byte_memset(unsigned char *out, unsigned len) {
+    case 1: copychunk(&out, &from);
     }
-
-    return chunk_memcpy(out, from, len);
+    len /= COPYCHUNK_UNROLL;
+    while (len > 0) {
+        if (COPYCHUNK_UNROLL >= 1) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 2) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 3) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 4) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 5) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 6) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 7) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 8) copychunk(&out, &from);
+        len--;
+    }
+#else
+    while (len-- > 0) {
+        storechunk(out, loadchunk(from));
+        out += INFFAST_CHUNKSIZE;
+        from += INFFAST_CHUNKSIZE;
+    }
+#endif
+    return out;
 }
 
-/* Byte by byte semantics: copy LEN bytes from FROM and write them to OUT. Return OUT + LEN. */
-static inline unsigned char *chunk_copy(unsigned char *out, unsigned char *from, int dist, unsigned len) {
-    if (len < sizeof(uint64_t)) {
-        if (dist > 0)
-            return set_bytes(out, from, dist, len);
+/*
+   Behave like chunkcopy_relaxed, but avoid writing beyond of legal output.
 
+   Accepts an additional pointer to the end of safe output.  A generic safe
+   copy would use (out + len), but it's normally the case that the end of the
+   output buffer is beyond the end of the current copy, and this can still be
+   exploited.
+ */
+static inline unsigned char *chunkcopy_safe(unsigned char *out, const unsigned char *from, unsigned len,
+                                           unsigned char *limit) {
+    Assert(out + len <= limit, "chunk copy exceeds safety limit");
+    if (limit - out < COPYCHUNKSIZE) {
+        /* Output buffer is too small to write a whole COPYCHUNKSIZE chunk, so
+           do things slowly and carefully. */
         return copy_bytes(out, from, len);
     }
+    return chunkcopy_relaxed(out, from, len);
+}
 
-    if (dist == 1)
-        return byte_memset(out, len);
-
-    if (dist > 0)
-        return chunk_memset(out, from, dist, len);
+/*
+   Perform short copies until distance can be rewritten as being at least
+   COPYCHUNKSIZE.
 
-    return chunk_memcpy(out, from, len);
+   This assumes that it's OK to overwrite at least the first 2*COPYCHUNKSIZE
+   bytes of output even if the copy is shorter than this.  This assumption
+   holds because inflate_fast() starts every iteration with at least 258 bytes
+   of output space available (258 being the maximum length output from a single
+   token; see inflate_fast()'s assumptions below).
+ */
+static inline unsigned char *chunkunroll_relaxed(unsigned char *out, unsigned *dist, unsigned *len) {
+    const unsigned char *from = out - *dist;
+    while (*dist < *len && *dist < COPYCHUNKSIZE) {
+        storechunk(out, loadchunk(from));
+        out += *dist;
+        *len -= *dist;
+        *dist += *dist;
+    }
+    return out;
 }
 
```

Comment:
In a revised version of this patch (internal review pending) I've separated `chunkunroll_*()` and `chunkunroll_lapped_*()`, such that the former takes `restrict` pointers which must _never_ overtlap, and the latter takes a pointer and a distance which _implicitly_ overlap.  This forces all calls to be explicit about whether or not they need the unrolling or `memset()`-like behaviour.

In the `chunkunroll_lapped_safe()` case, this also means internalising the guard for unrolling too close to the end of the buffer, so there's no confusion (or less confusion) about how much working room it needs.